### PR TITLE
Implement an SMB-Server on OS level

### LINKED ISSUE #1462: a script for configuring a SMB-server
Hey, I know that we have currently many things that are under construction and not done yet, but what do you think about a script for configuring a SMB-server? I've yesterday/tonight made one 😉
It works already quite nicely but I have to make some polishing like comments and wordings, have a look if travis works, and so on...

### PR DESCRIPTION
**Pull request by szaimen**
@enoch85 I would suggest that you test it out yourself.
I know it looks ugly especially with missing comments, etc (most likey travis will find a lot)
But in my testing it works quite nicely and is very feature rich ;)
Fix #1462

### CODE REVIEW

**Comment by szaimen** [comment]
You can download it with this command (just as a reminder for myself):
`wget https://raw.githubusercontent.com/nextcloud/vm/szaimen-patch-1/apps/smbserver.sh`

File: apps/smbserver.sh [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+    exit
+fi
+
+# Show explainer
+explainer_popup
+
+DIRECTORIES=$(find /mnt -maxdepth 3 -type d -not -path "/mnt/ncdata*" | grep -v "^/mnt$")
+if [ -z "$DIRECTORIES" ]
+then
+    msg_box "No directories found that can be used. Please make sure to mount them in '/mnt'."
```
Comment (szaimen) [inline apps/smbserver.sh:48]: To test this script, you can of course just create a folder inside /mnt

**Comment by enoch85** [comment]
Wow, someone's been busy! :D

**Comment by enoch85** [comment]
> Before running it, I would suggest creating a folder inside /mnt ;)

I think the script should take care of that. If no folder exists, then use `inputbox_flow` to let the user name it, and then create it.

**Comment by szaimen** [comment]
Have you tried it? :) What do you think?

> I think the script should take care of that. If no folder exists, then use `inputbox_flow` to let the user name it, and then create it.

I mean that would of course be possible. Shall I implement it this way? I just thought it would make sense to only use available folders...

**Comment by enoch85** [comment]
Just tested, the password prompt for the SMB-user is kind of funny, it never ends. 

1. The password has to contain three of this XXX
2. The password has to contain three of this xxx
3. The password has to contain three of this )(/
4. The password has to contain three of this 0987

I laughed out loud when I added the password, will it NEVER END?!?! hahaha

Also, I think that a lot of stuff can be made cleaner, and automated. I won't go in to details here though due to lack of time.

**Comment by szaimen** [comment]
> 3\. )(/

Those are not qualified quaracters ;)
only `.,#+-`
no other count to this

**Comment by szaimen** [comment]
> Also, I think that a lot of stuff can be made cleaner, and automated.

What thinks? A password generator maybe?

**Comment by szaimen** [comment]
> Just tested, the password prompt for the SMB-user is kind of funny, it never ends.
> 
> 1. The password has to contain three of this XXX
> 2. The password has to contain three of this xxx
> 3. The password has to contain three of this )(/
> 4. The password has to contain three of this 0987
> 
> I laughed out loud when I added the password, will it NEVER END?!?! hahaha

This was just fixed

**Comment by enoch85** [comment]
Regarding the password, it should be more obvious exactly what's required to pass the test for the password. I'm an experienced user, imagine those who aren't.

And, I'm not saying this will ever be merged, I'm just giving you advice here. :)

**Comment by szaimen** [comment]
> I'm just giving you advice here. :)

okay thanks

**Comment by enoch85** [comment]
We could merge this to a folder called "not supported", and link to it if users want it, and in that way don't include it in the released version but at the same time be tested by users. 

That way we could get feedback and maybe merge it into release once it's been tested for a long time.

Just an idea.

**Comment by szaimen** [comment]
> We could merge this to a folder called "not supported", and link to it if users want it, and in that way don't include it in the released version but at the same time be tested by users. 
> 
> That way we could get feedback and maybe merge it into release once it's been tested for a long time.
> 
> Just an idea.

Sounds good 👍

**Comment by enoch85** [comment]
OK, let's do that then. 

Please close this PR and merge it to the correct folder.

Fix travis and all that, when it's ready you tell me and I'll merge it. :+1: :100: :1st_place_medal:

**Comment by szaimen** [comment]
> OK, let's do that then.
> 
> Please close this PR and merge it to the correct folder.
> 
> Fix travis and all that, when it's ready you tell me and I'll merge it. 👍 💯 🥇

Sounds good. I would just work on this PR in this folder and move it when everything is fixed to the correct folder. Okay?

**Comment by enoch85** [comment]
> Okay?

Fine with me! :+1:

**Comment by szaimen** [comment]
@enoch85 I think this is ready :)

**Comment by szaimen** [comment]
image
funny 😄

**Comment by szaimen** [comment]
I would love if you would try it for yourself once more :)
I think I fixed most of the usability issues that you have reported.

**Comment by szaimen** [comment]
Pleas wait a second with merging. I want to add one thing...

**Comment by enoch85** [comment]
Just read the comment by @skjnldsv on the forum. 

Maybe this isn't such a bad idea after all... :thinking:

**Comment by skjnldsv** [comment]
> Just read the comment by @skjnldsv on the forum. 
> 
> Maybe this isn't such a bad idea after all... :thinking: 

The os level mount?

**Comment by enoch85** [comment]
> The os level mount?

Yes. :)

**Comment by enoch85** [comment]
@szaimen Great work, just tested. Much more user friendly now. Not 100% there yet, but a big difference from the first POC.

**Comment by szaimen** [comment]
> Much more user friendly now. Not 100% there yet, but a big difference from the first POC.

What is missing in your opinion? How can I make it 100% user friendly? 😉

File: not-supported/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+## Can I help?
+Yes, of course! :)
+Although some things in those scripts might not be 100% ready, we would love to hear your feedback regarding those.
+Feedback is especially welcome, if you would love to have some features that those script brings in the Release version of the NcVM.
+So please report back! 🚀
+
+### How to run this inside my NcVM?
+Currently there is only one script inside this folder. The command to get and run this script is the following:<br>
+`wget https://raw.githubusercontent.com/nextcloud/vm/master/not-supported/smbserver.sh && sudo bash smbserver.sh`
```
Comment (szaimen) [inline not-supported/README.md:12]: @enoch85 What do you think about this readme file?

Also, could this be added as a section to the main-readme file as well?

**Comment by szaimen** [comment]
@enoch85 I think this is ready if you have no additional comments

File: not-supported/README.md (hunk 2 above)
Comment (enoch85) [reply not-supported/README.md:12]: Nice README, but not for the main one. 

Maybe add a link to the folder under "Support the development"

**Comment by enoch85** [comment]
> 100% user friendly?

Well, a few things from the top of my head:
1. You don't tell the user what you expect from the password. Like it *need* to contain this to be valid.
2. You don't instruct on how to use the keys to select and choose
3. There's no "exit"
4. Wording can be improved. I notice you really like to use hyphens, and iften forget to capitalize. ;)

But 100% would only be the case if the script magically did everything without user input. :D

File: not-supported/README.md (hunk 2 above)
Comment (szaimen) [reply not-supported/README.md:12]: okay

Comment (szaimen) [reply not-supported/README.md:12]: done

**Comment by szaimen** [comment]
@enoch85 Thanks for the comment! I tried to address this with the last commit. At least point 1 and 3 should be fixed now.

> 2. You don't instruct on how to use the keys to select and choose

I would like to address this in a follow up for all whiptail --menu's I created #1477 for this

> 4. Wording can be improved. I notice you really like to use hyphens, and often forget to capitalize. ;)

I you have specific pointers, that would be really helpful for me, thanks!

> But 100% would only be the case if the script magically did everything without user input. :D

Maybe yes, but I intentionally wanted to avoid this to create a tool that can be used as a kind of full-featured SMB-server management-tool 😉

File: README.md [hunk 3]
```diff
@@ -33,6 +33,7 @@ Server installation. Simplified. :cloud:
 * Report [your issue](https://github.com/nextcloud/vm/issues/new)
 * Help us with [existing issues](https://github.com/nextcloud/vm/issues)
 * Write scripts so that the release process becomes automated with [Vagrant](https://www.vagrantup.com/docs/getting-started/), [Terraform](https://www.terraform.io/) or similar
+* Help us testing scripts, that aren't ready for the release version, yet. Please look at [this subfolder](https://github.com/nextcloud/vm/master/not-supported) for more infos.
```
Comment (enoch85) [inline README.md]: Released
Info
Remove the comma after scripts and yet

Comment (szaimen) [reply README.md]: tried to address this

File: README.md [hunk 4]
```diff
@@ -32,6 +32,7 @@ Server installation. Simplified. :cloud:
 * [Create a PR](https://help.github.com/articles/creating-a-pull-request/) and improve the code
 * Report [your issue](https://github.com/nextcloud/vm/issues/new)
 * Help us with [existing issues](https://github.com/nextcloud/vm/issues)
+* Test scripts that aren't ready for the Released version yet. Please look at [this subfolder](https://github.com/nextcloud/vm/master/not-supported) for further information.
```
Comment (szaimen) [inline README.md:35]: I hope this is better now?

Comment (enoch85) [reply README.md]: Fine with me! :)

**Comment by enoch85** [comment]
> If you have specific pointers, that would be really helpful for me, thanks!

Not right now sorry. 

From now on my time is very limited, so I won't be able to help out as much. :/

I'm fine with merging this so that users can test it.

**Comment by szaimen** [comment]
> From now on my time is very limited, so I won't be able to help out as much. :/

Ah okay?

> I'm fine with merging this so that users can test it.

Then please go ahead and merge it! 👍

**Comment by enoch85** [comment]
> Ah okay? How comes?

Starting my regular day-job now + daughter that needs food, care, love, and sleep when I get home.

**Comment by szaimen** [comment]
That's understandable 👍

**Pull request merged**